pkg/operator/controller/extension/virtualcluster/add.go

```diff
@@ -0,0 +1,89 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package virtualcluster
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/clock"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	operatorv1alpha1 "github.com/gardener/gardener/pkg/apis/operator/v1alpha1"
+	clientmapbuilder "github.com/gardener/gardener/pkg/client/kubernetes/clientmap/builder"
+	"github.com/gardener/gardener/pkg/controllerutils/mapper"
+	predicateutils "github.com/gardener/gardener/pkg/controllerutils/predicate"
+	operatorpredicate "github.com/gardener/gardener/pkg/operator/predicate"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "garden-extension-virtual-cluster"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager) error {
+	if r.RuntimeClient == nil {
+		r.RuntimeClient = mgr.GetClient()
+	}
+	if r.Clock == nil {
+		r.Clock = clock.RealClock{}
+	}
+	if r.Recorder == nil {
+		r.Recorder = mgr.GetEventRecorderFor(ControllerName + "-controller")
+	}
+	if r.GardenNamespace == "" {
+		r.GardenNamespace = v1beta1constants.GardenNamespace
+	}
+
+	if r.GardenClientMap == nil {
+		var err error
+		r.GardenClientMap, err = clientmapbuilder.
+			NewGardenClientMapBuilder().
+			WithRuntimeClient(r.RuntimeClient).
+			WithClientConnectionConfig(&r.Config.VirtualClientConnection).
+			WithGardenNamespace(r.GardenNamespace).
+			Build(mgr.GetLogger())
+		if err != nil {
+			return fmt.Errorf("failed to build garden ClientMap: %w", err)
+		}
+		if err := mgr.Add(r.GardenClientMap); err != nil {
+			return err
+		}
+	}
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&operatorv1alpha1.Extension{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: ptr.Deref(r.Config.Controllers.ExtensionVirtualCluster.ConcurrentSyncs, 0),
+		}).
+		Watches(
+			&operatorv1alpha1.Garden{},
+			mapper.EnqueueRequestsFrom(ctx, mgr.GetCache(), mapper.MapFunc(r.MapToAllExtensions), mapper.UpdateWithNew, mgr.GetLogger()),
+			builder.WithPredicates(predicate.Or(operatorpredicate.GardenCreatedOrReconciledSuccessfully(), predicateutils.ForEventTypes(predicateutils.Delete))),
```

Comment:
Wouldn't the `Extension` then be queued for reconciliation once the `Garden` finishes the it's own reconciliation since we watch on `Garden` events ?

The point of that additional watch is to trigger a reconciliation for all extension after a `Garden` resource has successfully finished it's reconcile (so that the virtual KAPI and GAPI are both ready to handle the requests). That being said, the reconciliation of the `Extension` is indeed delayed but without KAPI and GAPI it wouldn't work. We can refine the predicate more maybe for the very specific conditions

Reply:
I missed that the controller is primarily watching `Extension`s and reconciles them on generation change. It additionally watches the `Garden` and reconciles all `Extension`s` after the `Garden` was successfully reconciled. OK, so far so good.

However, can't even this lead to problems? Say, a new Gardener version is deployed and `Garden` is currently reconciled (which takes a while, maybe ETCD is rolled). In the meantime, a user bumped the version of an `Extension` object. Now, this controller will immediately roll this extension out, even though the `Garden` reconciliation has not finished.

I guess the reconciler should check if the `Garden` is currently in a reconciliation, and if yes, then requeue? WDYT?

Reply:
> I guess the reconciler should check if the Garden is currently in a reconciliation, and if yes, then requeue? WDYT?

This was the case many commits before. It changed after https://github.com/gardener/gardener/pull/10006#discussion_r1652637817 to "optimize" against a potential second reconcilation. 

Currently, in that scenario the worst case is that it would return an error because the vKAPI is not available. Though it's probably much uglier to do that, compared to checking the garden conditions for requeue as you proposed.

Reply:
Even if VKAPI is available - process-wise, we should wait for all this `Garden` reconciliation logic to finish before we update and rollout `Controller{Registration,Deployment}s`. So, I guess we cannot go around this check. I don't know whether there are other ways to optimize against additional reconciliations, but if not, we probably can live with it (the effect should be negligible).